Add `VectorWrapper` to improve usability from the C++ side

tobiasleibner opened the pull request:
With the wrapper we can wrap arbitrary types and still only need bindings for `VectorInterface`. 
Todo:

- [ ] Add tests for new code
- [x] Make wrapper work with types that are already derived from `VectorInterface`

tobiasleibner committed:
First draft of vector traits and wrapper

tobiasleibner committed:
Refine VectorWrapper and VectorTraits

tobiasleibner committed:
Add DynamicVector to test vectors derived from VectorInterface

tobiasleibner committed:
Fix msvc warning

ftschindler commented on src/nias_cpp/vector/traits.h:
```diff
@@ -0,0 +1,48 @@
+    } -> std::same_as<typename VectorTraits<VectorType>::ScalarType&>;
+    {
+        VectorTraits<VectorType>::const_get_(vec, 0)
+    } -> std::same_as<const typename VectorTraits<VectorType>::ScalarType&>;
```
we might weaken this to a
```suggestion
    } -> std::same_as<typename VectorTraits<VectorType>::ScalarType>;
```
as we might have getters that return by value, and we don't want to force having a ref to a temporary.

tobiasleibner replied:
I changed it to accept both (value and reference).

ftschindler commented:
Very nice! Things we might want to look into:
- [ ] `ConstVectorInterface` and `VectorInterface : ConstVectorInterface` as for the vector arrays
- [ ] how to hold the backend in the wrapper
  - const wrapper holding only ref, only implementing `ConstVectorInterface`
  - wrapper holding mutable ref, implementing `VectorInterface`
  - wrapper doing a copy? Do we want that ever, probably not?
  - `ConstVectorWrapperTraits` and `VectorWrapperTraits : ConstVectorWrapperTraits`
- [ ] the `VectorTraits` are specific to the nias wrapper, and will likely not be reused for other contexts. So we can
  - drop the trailing `_`?
  - rename them to `VectorWrapperTraits`?
  - rename `has_vector_traits` -> `wrappable_vector`? (one or two `p`? ;P)

tobiasleibner commented:
> * wrapper doing a copy? Do we want that ever, probably not?

Maybe to make wrappers copyable? So if we have a `ConstRefWrapper<VectorType>` and want to make a copy, we need a `OwningWrapper<VectorType>` to hold the copied vector.

ftschindler commented:
> Very nice! Things we might want to look into:
> 
>     * [ ]  `ConstVectorInterface` and `VectorInterface : ConstVectorInterface` as for the vector arrays
> 
>     * [ ]  how to hold the backend in the wrapper
>       
>       * const wrapper holding only ref, only implementing `ConstVectorInterface`
>       * wrapper holding mutable ref, implementing `VectorInterface`
>       * wrapper doing a copy? Do we want that ever, probably not?
>       * `ConstVectorWrapperTraits` and `VectorWrapperTraits : ConstVectorWrapperTraits`
> 
>     * [ ]  the `VectorTraits` are specific to the nias wrapper, and will likely not be reused for other contexts. So we can
>       
>       * drop the trailing `_`?
>       * rename them to `VectorWrapperTraits`?
>       * rename `has_vector_traits` -> `wrappable_vector`? (one or two `p`? ;P)

- `ConstVectorWrapper` throws in mutable methods
- `VectorReferenceWrapper : ConstVectorWrapper`
- `OwningVectorWrapper : VectorReferenceWrapper`
- just one traits, because copy returns `OwningVectorWrapper`

On ListVectorArray:
- use std::variant to allow all three kinds of wrappers internally

tobiasleibner commented:
> On ListVectorArray:
> 
>     * use std::variant to allow all three kinds of wrappers internally

We might also want to look into the [proxy library](https://github.com/microsoft/proxy).

tobiasleibner committed:
Add some comments

tobiasleibner committed:
Update pre-commit hooks

tobiasleibner committed:
Add dependabot config

tobiasleibner committed:
Drop trailing _ for VectorTraits members

tobiasleibner committed:
Update uv and pybind11

tobiasleibner committed:
Allow to have owning and non-owning vector wrappers

tobiasleibner committed:
Rename VectorTraits and has_vector_traits

tobiasleibner committed:
Use C++23

tobiasleibner committed:
Fix forwarding errors

tobiasleibner committed:
Update pyproject.toml

tobiasleibner committed:
Update pre-commit hooks

tobiasleibner committed:
Update dependency handling

tobiasleibner committed:
Add boost dependency to fix Python handling

Before, when the first Python interpreter on the PATH had a different
version than the one we linked, the test executables failed at startup.
boost now has to be available during build time. I tried to download it
automatically when it is not available (as we do for the other
dependencies) but that proved to be very complicated (linking was no
problem, but installing nias_cpp afterwards was a big problem because we
then also have to export all of the boost targets manually etc.)

tobiasleibner committed:
Add forward_like implementation

tobiasleibner committed:
Update workflows

tobiasleibner committed:
Fix clang-tidy

tobiasleibner committed:
More boost installations

tobiasleibner committed:
Fix test project

tobiasleibner committed:
Add missing find_package call

tobiasleibner committed:
Fix more clang-tidy complaints

tobiasleibner committed:
Fix python handling

tobiasleibner committed:
Another try to fix boost installation on Windows

tobiasleibner committed:
Explicitly use boost process v2

tobiasleibner committed:
Yet another attempt to fix Boost on Windows

tobiasleibner committed:
Improve boost handling

tobiasleibner committed:
Debug output for pybind11 setup

tobiasleibner committed:
Do not install dependencies from config

tobiasleibner committed:
Update VectorWrapperTraits documentation

tobiasleibner committed:
Try to fix downstream inclusion workflow

tobiasleibner committed:
Fix python installation

tobiasleibner committed:
Update pre-commit hooks

tobiasleibner committed:
Another attempt to fix Python detection

tobiasleibner committed:
Move Python handling to nias_cpp-config.cmake

tobiasleibner committed:
Fix virtualenv usage

tobiasleibner committed:
A few more fixes

tobiasleibner committed:
Execute test project

tobiasleibner committed:
Fix export header

tobiasleibner committed:
Improve PYTHONHOME search

tobiasleibner committed:
Try to fix virtualenv downstream inclusion

tobiasleibner committed:
Try to fix installation paths

tobiasleibner committed:
Make install directory configurable

tobiasleibner committed:
Try to fix vcpkg installation

tobiasleibner committed:
Another try to fix vcpkg port

tobiasleibner committed:
Fix cmake dir output

tobiasleibner committed:
Fix export header

tobiasleibner committed:
Fix undefined symbol errors on Windows

tobiasleibner committed:
Yet another try to make all platforms happy

tobiasleibner committed:
Enable warnings-as-errors for test projects

tobiasleibner committed:
Export methods instead of whole class

tobiasleibner committed:
Another try to fix visibility issues

tobiasleibner committed:
Use pimpl idiom to fix visibility issues

tobiasleibner committed:
Reenable hidden visibility preset

tobiasleibner committed:
Pass python location from CMake to C++

tobiasleibner committed:
Try to remove check

tobiasleibner committed:
Use PyConfig.home instead of deprecated Py_SetPythonHome

tobiasleibner committed:
Remove boost dependency

tobiasleibner committed:
Fix clang-tidy

tobiasleibner committed:
Add CodeQL file to gitignore

tobiasleibner committed:
Remove explicit Python lib linking

tobiasleibner committed:
Fix compilation on Windows

tobiasleibner committed:
Always build in parallel

tobiasleibner committed:
Increase version to v0.1.3

tobiasleibner committed:
Also run clang-tidy in parallel

tobiasleibner committed:
Add missing return statement

tobiasleibner committed:
Increase version

tobiasleibner committed:
Clear existing environment before creating new one

tobiasleibner committed:
Apply updated clang-tidy config

tobiasleibner committed:
Update version to 0.1.5
